ZKProof/Plonkish/Defs.lean

```diff
@@ -0,0 +1,237 @@
+import Mathlib.Data.Rel
+import Mathlib.Data.Fin.Basic
+import Mathlib.Logic.Basic
+import Mathlib.Logic.Function.Defs
+import Mathlib.Algebra.Group.Defs
+import Mathlib.Algebra.Field.Defs
+import Mathlib.Tactic
+
+import ZKProof.Relations
+
+variable (F : Type) [Field F]
+
+
+/-- The location of a witness entry. -/
+structure Location (m n : ℕ) where
+  /-- Column number. -/
+  i : Fin m
+  /-- Row number. -/
+  j : Fin n
+deriving DecidableEq
+
+/-- Utility function to construct a `Location`. -/
+def entry {m n : ℕ} (i : Fin m) (j : Fin n) : Location m n := { i := i, j := j }
+
+/-- An abstract Plonkish circuit. -/
+structure AbstractCircuit where
+  /-- Length of the instance vector. -/
+  t : ℕ
+
+  /-- Number of columns for the witness matrix. -/
+  m : ℕ
+  m_pos : m > 0 := by simp
+  /-- Number of rows for the witness matrix. -/
+  n : ℕ
+  n_pos : n > 0 := by simp
+
+  /-- An equivalence relation on witness entries. -/
+  E (e e' : Location m n) : Prop
+  Equivalence_E : Equivalence E
+
+  /-- Where each instance vector entry corresponds to in the witness matrix. -/
+  S : Vector (Location m n) t
+
+  /-- The number of columns that are fixed. -/
+  m_f : Fin (m+1)
+  /-- The fixed content of the first `m_f` columns. -/
+  f (e : Location m n) (is_fixed_col : e.i < m_f) : F
+
+  /-- The number of custom gates. The default is 0, in which case `p` and `CUS` need not be provided. -/
+  U : ℕ := 0
+  /-- Multivariate polynomials for custom gates. -/
+  p (u : Fin U) : MvPolynomial (Fin m) F := by intro u; exact Fin.elim0 u
+  /-- Rows on which the custom polynomials `p u` are constrained to evaluate to 0. -/
+  CUS (u : Fin U) : Set (Fin n) := by intro u; exact Fin.elim0 u
+
+  /-- The number of lookup tables. The default is 0, in which case `L`, `TAB`, `q` and `LOOK` need not be provided. -/
+  V : ℕ := 0
+  /-- The number of table columns in the lookup table with index `v`, `TAB v`. -/
+  L (v : Fin V) : ℕ := by intro v; exact Fin.elim0 v
+  /-- Lookup tables `TAB v`, each with a number of tuples in `F^{L v}`. -/
+  TAB (v : Fin V) : Set (Vector F (L v)) := by intro v; exact Fin.elim0 v
+  /-- Scaling multivariate polynomials for lookup tables. -/
+  q (v : Fin V) : Vector (MvPolynomial (Fin m) F) (L v) := by intro v; exact Fin.elim0 v
+  /-- Rows on which the scaling polynomials `q v s` evaluate to some tuple in `TAB v`. -/
+  LOOK (v : Fin V) : Set (Fin n) := by intro v; exact Fin.elim0 v
+
+namespace AbstractCircuit
+variable (C : AbstractCircuit F)
+
+/-- The witness entry type for an `AbstractCircuit`. -/
+abbrev Entry := Location C.m C.n
```

Comment:
Replace Entry by WitnessPos

Reply:
Entry is the term used in the doc (e.g. see https://github.com/zkpstandard/wg-plonkish/blob/main/src/relation.md#copy-constraints).